### PR TITLE
examples: Variational Autoencoder on MNIST

### PR DESCRIPTION
**Pull request by alicanb**
add variational autoencoder on MNIST dataset as example. It doesn't train efficiently with gradient descent so I also implemented rmsprop as well. Here's how the digits look after 10 epochs: 
res

### CODE REVIEW

**Comment by lenck** [comment]
Hi, I've merged it with the latest devel to a branch `alicanb-vae`. There still need to be done some changes (e.g. have vl_simplenn interface first, as is usual for the rest of the functions, write tests, write some documentation).
I wanted to ask you, would you have some more details for the vae loss implementation? E.g. justification for the numerical tricks there etc.

Thanks for the Pull Request btw! :)

**Comment by alicanb** [comment]
Thanks for the merge :) I think it will be hard to write the vl_simplenn interface since the architecture is not a linear chain (Gaussian has 2 inputs, VaeLoss has 4).  

I would be happy to provide additional documentation, I will follow up with another PR to the new branch. The only trick I applied is the epsilon in single format for stable calculation of cross entropy (which I think should be an option in ```vl_nnloss```), the loss formulas are from [Auto-encoding Variational Bayes](https://arxiv.org/abs/1312.6114). About the details of the implementation, I ported from [Keras](https://github.com/fchollet/keras/blob/master/examples/variational_autoencoder.py).
